### PR TITLE
Ipmi reboot

### PR DESCRIPTION
**Pull request by twaldrop**
Move from manual reboot for testing failover to hard reboot via IPMI.

### CODE REVIEW

**Comment by j2sol** [comment]
:+1:

**Comment by j2sol** [comment]
I'd like to test this play out on trustytest, then will merge.

**Comment by dlundquist** [comment]
I'm still opposed to adding access to the IPMI interface into Ursula. This restricts to only testing machines with IPMI, requires the host we invoke the playbook to have privileged access to the IPMI network, and there doesn't seem to be a strong reason to use it over the magic sysrq interface i.e. using 'b' instead of 'c'.

**Comment by j2sol** [comment]
My goal is to initiate a hard reset, so that the host automatically comes back and does not require manual mucking with IPMI or remote power controls. If the sysrq method can do a reboot vs a panic, I'm 100% fine with that. I'd just like it to be scripted rather than manually done. Is that acceptable?

**Comment by dlundquist** [comment]
@j2sol I agree on a scripted test, my concern is adding IPMI as a requirement for performing the test which will prevent us from testing on VMs, servers without IPMI or cases where the IPMI network isn't reachable inband from the host we initiate the test run.

File: branch-validation.yml [hunk 1]
```diff
@@ -171,13 +171,14 @@
...
-    # This is manual for now, not keen on automated kernel panic
+    # initiate hard reboot on the current master controller
     - name: initiate a failover on controller 0
-      pause: prompt="On ucarp {{ inventory_hostname }} $ echo c > /proc/sysrq-trigger (press enter when done)"
+      shell: echo 'b' | tee /proc/sysrq-trigger
       when: master is defined
 
     - name: initiate a failover on controller 1
-      pause: prompt="On ucarp {{ groups['controller'][1] }} $ echo c > /proc/sysrq-trigger (press enter when done)"
+      delegate_to: "{{ groups['controller'][1] }}"
```
Comment (j2sol) [inline branch-validation.yml]: Minor nit. The delegate_to should be the last key on the task hash, just a style thing.

File: branch-validation.yml [hunk 2]
```diff
@@ -229,13 +227,13 @@
       when: master is not defined
 
     - name: floating ips are reachable
-      command: ping -c 5 {{ item.public_ip }}
+      command: ping -c 60 {{ item.public_ip }}
```
Comment (j2sol) [inline branch-validation.yml]: Whoh, why are we pinging 60 times? Shouldn't 5 successful pings be sufficient?

File: branch-validation.yml [hunk 3]
```diff
@@ -229,13 +227,13 @@
...
       with_items: fip_controllers.results + fip_compute.results
       changed_when: false
       delegate_to: localhost
       register: last_ping
       until: comp_ping|success
-      retries: 5
+      retries: 60
```
Comment (j2sol) [inline branch-validation.yml]: Just to be sure, this is going to basically be a timeout of 60*10, or 600 seconds, which is 10 minutes. Is 10 minutes the time you're looking for here?

**Comment by j2sol** [comment]
Some comments in code, and unfortunately it needs a rebase.

File: branch-validation.yml (hunk 2 above)
Comment (twaldrop) [inline branch-validation.yml]: Thanks for catching that 5 pings with 5 retries was consistently catching the server before it was actually back up, when it was responding to ping, but not ssh.  Decreasing the values and rebasing

File: branch-validation.yml [hunk 4]
```diff
@@ -183,26 +183,30 @@
...
 
     - name: wait for the panic
       wait_for: host={{ groups['controller'][1] }} port=22 state=stopped
                 delay=10 timeout=30
-      delegate_to: localhost
       when: master is not defined
+      delegate_to: localhost
+
+    - name: wait for the node to boot
+      pause: minutes=3
```
Comment (j2sol) [inline branch-validation.yml]: oh, I see what's going on here. The wait_for is already waiting for ssh to stop responding, but even then ping would respond. I get what you're saying. 3 minutes here might be too long, is there a chance the host would be already coming back up by then?

Comment (twaldrop) [inline branch-validation.yml]: That makes a lot of sense.  30 seconds should be sufficient.

**Comment by j2sol** [comment]
retest

**Pull request merged**